File: src/server_common/mpwp_protocol.py
import uuid

# MSG PARTS FOR FLEXIBLE INDEXING

MSG_VERSION = 0
MSG_STATUS = 1
MSG_TO = 2
MSG_FROM = 3
MSG_MSGNUM = 4
MSG_TYPE = 5
MSG_CONTENT = 6

# VERSION TAG

VERSION = b'mpwp_v1.0'

# STATUS CODES

STATUS_DATA = b'100'
STATUS_PING = b'101'

STATUS_CONNECT = b'200'
STATUS_CONNECT_OK = b'201'
STATUS_DISCONNECT = b'202'
STATUS_REGISTER = b'203'
STATUS_REGISTER_OK = b'204'
STATUS_RESYNC = b'205'
STATUS_RESYNC_OK = b'206'

STATUS_LOG = b'300'

STATUS_SERVER_ERROR = b'500'
STATUS_VERSION_MISMATCH_ERROR = b'501'
STATUS_INCORRECT_ID_ERROR = b'502'

# SERVER IDS

MATCHMAKER_ID = b'0'
GAME_MANAGER_ID = b'1'
CLIENT_HANDLER_ID = b'2'
LOGGER_ID = b'3'

# GAME MSG TYPES

GAME_CREATE = b'0'
GAME_STATE = b'1'
GAME_INPUT = b'2'

# MATCHMAKER MSG TYPES

MATCHMAKER_QUEUE = b'0'
MATCHMAKER_DEQUEUE = b'1'
MATCHMAKER_FOUND = b'2'
MATCHMAKER_ACCEPT = b'3'
MATCHMAKER_DECLINE = b'4'
MATCHMAKER_LOADING = b'5'
MATCHMAKER_LAUNCH = b'6'
MATCHMAKER_SYNC = b'7'


def get_uuid():
    return ("{}".format(uuid.uuid1())).encode()


def get_mpwp_packet(STATUS, TO, FROM, MSGNUM, TYPE):
    return [VERSION, STATUS, TO, FROM, MSGNUM, TYPE]


def get_mpwp_status_packet(STATUS, TO, FROM):
    return [VERSION, STATUS, TO, FROM]


def get_mpwp_content_packet(TO, FROM, MSGNUM, TYPE, CONTENT=None):
    packet = get_mpwp_packet(STATUS_DATA, TO, FROM, MSGNUM, TYPE)
    if CONTENT:
        if type(CONTENT) == list:
            packet = packet + CONTENT
        elif type(CONTENT) is bytes:
            packet.append(CONTENT)
        else:
            raise TypeError
    return packet


def get_log_packet(FROM, MSGNUM, TYPE, MSG=b''):
    packet = [VERSION, STATUS_LOG, LOGGER_ID, FROM, MSGNUM, TYPE, MSG]
    return packet


def msg_content(msg):
    return msg[MSG_CONTENT:]

File: src/client/lib/io/mpwp_protocol.dart
import 'package:client/io/mpwp_message.dart';

class MPWPProtocol
{
  static final int MSG_VERSION = 0;
  static final int MSG_STATUS = 1;
  static final int MSG_TO = 2;
  static final int MSG_FROM = 3;
  static final int MSG_MSGNUM = 4;
  static final int MSG_TYPE = 5;
  static final int MSG_CONTENT = 6;

  static final String VERSION = "mpwp_v1.0";

  static final String STATUS_DATA = "100";
  static final String STATUS_PING = "101";

  static final String STATUS_CONNECT = "200";
  static final String STATUS_CONNECT_OK = "201";
  static final String STATUS_DISCONNECT = "202";
  static final String STATUS_REGISTER = "203";
  static final String STATUS_REGISTER_OK = "204";
  static final String STATUS_RESYNC = "205";
  static final String STATUS_RESYNC_OK = "206";

  static final String STATUS_LOG = "300";

  static final String STATUS_SERVER_ERROR = "500";
  static final String STATUS_VERSION_MISMATCH_ERROR = "501";
  static final String STATUS_INCORRECT_ID_ERROR = "502";

  static final String MATCHMAKER_ID = "0";
  static final String GAME_MANAGER_ID = "1";
  static final String CLIENT_HANDLER_ID = "2";
  static final String LOGGER_ID = "3";


  static final String GAME_CREATE = "0";
  static final String GAME_STATE = "1";
  static final String GAME_INPUT = "2";

  static final String MATCHMAKER_ENQUEUE = "0";
  static final String MATCHMAKER_DEQUEUE = "1";
  static final String MATCHMAKER_FOUND = "2";
  static final String MATCHMAKER_ACCEPT = "3";
  static final String MATCHMAKER_DECLINE = "4";
  static final String MATCHMAKER_LOADING = "5";
  static final String MATCHMAKER_LAUNCH = "6";
  static final String MATCHMAKER_SYNC = "7";

  static MPWPMessage getMessage(String status, String to, String from, int msgnum, String type, [content=null])
  {
    List messageParts = [VERSION, status, to, from, msgnum.toString(), type];

    if(content != null)
    {
      if(content is List)
      {
        messageParts.addAll(content);
      }
      else if(content is String)
      {
        messageParts.add(content);
      }
    }

    return new MPWPMessage(messageParts);
  }

  static MPWPMessage getStatusMessage(String status, String to, String from)
  {
    List messageParts = [VERSION, status, to, from];
    return new MPWPMessage(messageParts);
  }
}